rgw/cloud-restore: handle versioned objects

soumyakoduri opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

thotz commented on src/rgw/driver/rados/rgw_lc_tier.cc:
```diff
@@ -323,7 +323,21 @@ int rgw_cloud_tier_get_object(RGWLCCloudTierCtx& tier_ctx, bool head,
+      std::replace(s1.begin(), s1.end(), '_', '-');
+
+      /* copy s3cmd attrs */
+      std::string a_p = "x-amz-meta-s3cmd-attrs";
```
Is this same for aws cli as well?

thotz commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5240,6 +5239,14 @@ int RGWRados::restore_obj_from_cloud(RGWLCCloudTierCtx& tier_ctx,
+    const auto aiter = attrs.find("x-amz-meta-rgwx-versioned-epoch");
+    if (aiter != std::end(attrs)) {
+      attrs[RGW_ATTR_RESTORE_VERSIONED_EPOCH] = aiter->second;
+      uint64_t versioned_epoch = stoi(rgw_bl_str(aiter->second));
```
do we need use this variable any where?

thotz commented on src/rgw/driver/rados/rgw_sal_rados.cc:
```diff
@@ -2750,12 +2750,19 @@ int RadosObject::handle_obj_expiry(const DoutPrefixProvider* dpp, optional_yield
+          auto v_iter = attrs.find(RGW_ATTR_RESTORE_VERSIONED_EPOCH);
+          if (v_iter != attrs.end()) {
+            uint64_t versioned_epoch = stoi(rgw_bl_str(v_iter->second));
+            obj_op.meta.olh_epoch = versioned_epoch;
```
olh_epoch passed to restore_obj_from_cloud is not used anywhere right? Do we still need that variable

soumyakoduri committed:
rgw/cloud-restore: Handle versioned objects

Use `key.instance` to restore versioned objects. Non-current objects
should remain non-current. This is done by reading `olh_epoch` from
the transitioned object and saving it in the new attr
"RGW_ATTR_RESTORE_VERSIONED_EPOCH" to be used while updating bi.

Also retain object attrs while restoring the objects.

Signed-off-by: Soumya Koduri <skoduri@redhat.com>
